cleanrl/rnd_ppo.py

```diff
@@ -0,0 +1,1079 @@
+
+    for i_epoch_pi in range(args.update_epochs):
+        target_agent.load_state_dict(agent.state_dict())
+        np.random.shuffle(inds)
+        for start in range(0, args.batch_size, args.minibatch_size):
+            end = start + args.minibatch_size
+            minibatch_ind = inds[start:end]
+            predict_next_state_feature, target_next_state_feature = rnd_model(rnd_next_obs[minibatch_ind])
+            forward_loss = forward_mse(predict_next_state_feature, target_next_state_feature.detach()).mean(-1)
+            
+            mask = torch.rand(len(forward_loss)).to(device)
+            mask = (mask < args.update_proportion).type(torch.FloatTensor).to(device)
+            forward_loss = (forward_loss * mask).sum() / torch.max(mask.sum(), torch.Tensor([1]).to(device))
+            del mask
+            mb_advantages = b_advantages[minibatch_ind]
+            if args.norm_adv:
+                mb_advantages = (mb_advantages - mb_advantages.mean()) / (mb_advantages.std() + 1e-8)
+
+            _, newlogproba, entropy = agent.get_action(b_obs[minibatch_ind], b_actions.long()[minibatch_ind])
+            ratio = (newlogproba - b_logprobs[minibatch_ind]).exp()
+
+            # Stats
+            approx_kl = (b_logprobs[minibatch_ind] - newlogproba).mean()
+
+            # Policy loss
+            pg_loss1 = -mb_advantages * ratio
+            pg_loss2 = -mb_advantages * torch.clamp(ratio, 1-args.clip_coef, 1+args.clip_coef)
+            pg_loss = torch.max(pg_loss1, pg_loss2).mean()
+            entropy_loss = entropy.mean()
+
+            # Value loss
+            new_ext_values, new_int_values = agent.get_value(b_obs[minibatch_ind])
+            new_ext_values, new_int_values = new_ext_values.view(-1), new_int_values.view(-1)
+            if args.clip_vloss:
+                ext_v_loss_unclipped = ((new_ext_values - b_ext_returns[minibatch_ind]) ** 2)
+                ext_v_clipped = b_ext_values[minibatch_ind] + torch.clamp(new_ext_values - b_ext_values[minibatch_ind], -args.clip_coef, args.clip_coef)
+                ext_v_loss_clipped = (ext_v_clipped - b_ext_returns[minibatch_ind])**2
+                ext_v_loss_max = torch.max(ext_v_loss_unclipped, ext_v_loss_clipped)
+                ext_v_loss = 0.5 * ext_v_loss_max.mean()
+            else:
+                ext_v_loss = 0.5 *((new_ext_values - b_ext_returns[minibatch_ind]) ** 2).mean()
+
+            int_v_loss = 0.5 *((new_int_values - b_int_returns[minibatch_ind]) ** 2).mean()
+            v_loss = ext_v_loss + int_v_loss
+
+            optimizer.zero_grad()
+            loss = pg_loss - args.ent_coef * entropy_loss + v_loss * args.vf_coef + forward_loss
+
+            loss.backward()
+            nn.utils.clip_grad_norm_(list(agent.parameters()) + list(rnd_model.predictor.parameters()), args.max_grad_norm)
+            optimizer.step()
+
+            del loss
```

Comment:
I assume this will be taken care for by the garbage collector. I suggest removing it.